core/consensus/babe/src/epoch_changes.rs

```diff
@@ -0,0 +1,224 @@
+use parking_lot::{Mutex, MutexGuard};
+use sr_primitives::traits::{Block as BlockT, NumberFor};
+use codec::{Encode, Decode};
+use client::error::Error as ClientError;
+use client::utils as client_utils;
+use client::blockchain::HeaderBackend;
+use primitives::H256;
+
+/// A builder for `is_descendent_of` functions.
+pub trait IsDescendentOfBuilder<Block: BlockT> {
+	/// The error returned by the function.
+	type Error: std::error::Error;
+	/// A function that can tell you if the second parameter is a descendent of
+	/// the first.
+	type IsDescendentOf: Fn(&Block::Hash, &Block::Hash) -> Result<bool, Self::Error>;
+
+	/// Build an `is_descendent_of` function.
+	///
+	/// The `current` parameter can be `Some` with the details a fresh block whose
+	/// details aren't yet stored, but its parent is.
+	///
+	/// The format of `current` when `Some` is `(current, current_parent)`.
+	fn build_is_descendent_of(&self, current: Option<(Block::Hash, Block::Hash)>)
+		-> Self::IsDescendentOf;
+}
+
+// TODO: relying on Hash = H256 is awful.
+// https://github.com/paritytech/substrate/issues/3624
+impl<'a, Block: BlockT<Hash=H256>, T> IsDescendentOfBuilder<Block> for &'a T
+	where T: HeaderBackend<Block>
+{
+	type Error = ClientError;
+	type IsDescendentOf = Box<dyn Fn(&H256, &H256) -> Result<bool, ClientError> + 'a>;
+
+	fn build_is_descendent_of(&self, current: Option<(H256, H256)>)
+		-> Self::IsDescendentOf
+	{
+		Box::new(client_utils::is_descendent_of(*self, current))
+	}
+}
+
+/// Tree of all epoch changes across all *seen* forks. Data stored in tree is
+/// the hash and block number of the block signaling the epoch change, and the
+/// epoch that was signalled at that block.
+#[derive(Clone, Encode, Decode)]
+pub struct EpochChanges<Block: BlockT> {
+	inner: ForkTree<Block::Hash, NumberFor<Block>, Epoch>,
+}
+
+// create a fake header hash which hasn't been included in the chain.
+fn fake_head_hash<H: AsRef<[u8]> + AsMut<[u8]> + Clone>(parent_hash: &H) -> H {
+	let mut h = parent_hash.clone();
+	// dirty trick: flip the first bit of the parent hash to create a hash
+	// which has not been in the chain before (assuming a strong hash function).
+	h.as_mut()[0] ^= 0b10000000;
+	h
+}
+
+impl<Block: BlockT> EpochChanges<Block> {
+	/// Create a new epoch-change tracker.
+	fn new() -> Self {
+		EpochChanges { inner: ForkTree::new() }
+	}
+
+	/// Prune out finalized epochs, except for the ancestor of the finalized block.
+	pub fn prune_finalized<D: IsDescendentOfBuilder<Block>>(
+		&mut self,
+		descendent_of_builder: D,
+		_hash: &Block::Hash,
+		_number: NumberFor<Block>,
+	) -> Result<(), fork_tree::Error<D::Error>> {
+		let _is_descendent_of = descendent_of_builder
+			.build_is_descendent_of(None);
+
+		// TODO:
+		// prune any epochs which could not be _live_ as of the children of the
+		// finalized block.
+		// i.e. re-root the fork tree to the earliest ancestor of (hash, number)
+		// where epoch.start_slot + epoch.duration >= slot(hash)
```

Comment:
`self.inner.prune(hash, number, is_descendent_of)` should work, but I'll verify again before merging this PR.